RBD Regional DR: Implement clusterID mapping

Madhu-1 opened the pull request:
Whenever Ceph-CSI receives a CSI/Replication request it will first decode the volumeHandle and try to get the required OMAP details if it is not able to retrieve receives a `Not Found` error message and Ceph-CSI will check for the clusterID mapping. If the old volumeID `0001-00013-site1-storage-0000000000000001-b0285c97-a0ce-11eb-8c66-0242ac110002`
contains the `site1-storage` as the clusterID, now Ceph-CSI will look for the corresponding clusterID `site2-storage` from the clusterID mapping file. If the clusterID mapping is found now Ceph-CSI will look for the poolID mapping ie mapping between `1` and `2`. Example:- pool with name exists on both the clusters with different IDs Replicapool with ID `1` on site1 and Replicapool
with ID `2` on site2.  After getting the required mapping Ceph-CSI has the required information to get more details from the rados OMAP. If we have multiple clusterID mapping it will loop through all the mapping and checks the corresponding pool to get the OMAP data. If the clusterID mapping does not exist Ceph-CSI will return a `Not Found` error message to the caller.

Fixes: #2118
Depends-on: #2292

## Note to Reviewers

Up for initial review

## TODO:

- [x] Add example configmap yaml
- [x] Update required deployment templates to mount the configmap as volume
- [x] E2E testing

Madhu-1 committed:
util: add helper function to read clusterID mapping

added helper function to read the clusterID mapping
from the mounted file.

The clusterID mapping contains below mappings
* ClusterID mappings (to cluster to which we are failingover
and from which cluster failover happened)
* RBD PoolID mapping of between the clusters.
* CephFS FscID mapping between the clusters.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
rbd: read clusterID and PoolID from mapping

Whenever Ceph-CSI receives a CSI/Replication
request it will first decode the
volumeHandle and try to get the required
OMAP details if it is not able to
retrieve, receives a `Not Found` error
message and Ceph-CSI will check for the
clusterID mapping. If the old volumeID
`0001-00013-site1-storage-0000000000000001
-b0285c97-a0ce-11eb-8c66-0242ac110002`
contains the `site1-storage` as the clusterID,
now Ceph-CSI will look for the corresponding
clusterID `site2-storage` from the above configmap.
If the clusterID mapping is found now Ceph-CSI
will look for the poolID mapping ie mapping between
`1` and `2`. Example:- pool with name exists on
both the clusters with different ID's Replicapool
with ID `1` on site1 and Replicapool with ID `2`
on site2. After getting the required mapping Ceph-CSI
has the required information to get more details
from the rados OMAP. If we have multiple clusterID mapping
it will loop through all the mapping and checks the
corresponding pool to get the OMAP data. If the clusterID
mapping does not exist Ceph-CSI will return an `Not Found`
error message to the caller.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -1020,12 +1013,64 @@ func genVolFromVolID(
 	}
 
+	// Check clusterID mapping exists
+	mapping, err := util.GetClusterMappingInfo(vi.ClusterID)
```
This should probably be in a helper function.

nixpanic commented on internal/util/cluster_mapping.go:
```diff
@@ -0,0 +1,111 @@
+	CephFSpoolIDMappingInfo []map[string]string `json:"CephFSPoolMapping"`
+}
+
+// Expected JSON structure in the passed in config file is,
```
This is under discussion in #2292, maybe mark this PR as a draft?

I have added `Depends-on:` in the PR description, Mergify will now only merge this PR after the other is merged (new feature).

Madhu-1 committed:
deploy: add template changes for mapping

added template changes for the clusterID and
poolID,fsID mapping details for the pod templates.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

nixpanic commented on internal/util/cluster_mapping.go:
```diff
@@ -0,0 +1,117 @@
+)
+
+// clusterMappingConfigFile is the location of the cluster mapping config file.
+var clusterMappingConfigFile = "/etc/ceph-csi-config/cluster-mapping.json"
```
make this a const?

Madhu-1 replied:
added this as var to modify this on in go tests and verify different combinations. if we make it const the value cannot be changed and even we don't have access to create files in the `/etc` folder in go tests.

nixpanic commented on internal/util/cluster_mapping.go:
```diff
@@ -0,0 +1,117 @@
+// ...
+// }]
+
+func readClusterMappingInfo(pathToMappingConfig string) (*[]ClusterMappingInfo, error) {
```
does this really need to be a configurable path?

Madhu-1 replied:
followed the pattern same in csiconfig.go. this can be changed also

Madhu-1 replied:
updated now not to take this as input param

nixpanic commented on internal/util/cluster_mapping.go:
```diff
@@ -0,0 +1,117 @@
+
+func readClusterMappingInfo(pathToMappingConfig string) (*[]ClusterMappingInfo, error) {
+	var info []ClusterMappingInfo
+	// #nosec
```
`#nosec` is only needed it the path is non-const, consider making it a const.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -1020,12 +1013,37 @@ func genVolFromVolID(
+			return rbdVol, vErr
+		}
+	}
+	// TODO: remove extracting volumeID from PV annotations.
```
are these PV annotations documented somewhere? is it because the mapping might not be sufficient?

Madhu-1 replied:
No, it's not documented. This was the earlier logic to identify the mapping which is added by the controller we start in the provisioner pod.  with new changes mapping is sufficient. Will remove logic of adding/retrieving PV annotation to identify the mapping in next release. will open an issue for the same.

nixpanic replied:
@Madhu-1 don't forget to open an issue about this!

Madhu-1 replied:
https://github.com/ceph/ceph-csi/issues/2328 here it is

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -1043,15 +1061,89 @@ func genVolFromVolID(
+	return vol, nil
+}
+
+func getMappedID(key, value, id string) string {
```
A comment would help explaining what this functions purpose is

nixpanic commented on examples/csi-config-map-sample.yaml:
```diff
@@ -39,5 +48,20 @@ data:
+  cluster-mapping.json: |-
+    [{
+      "clusterIDMapping": {
+      "clusterID on site1: "clusterID on site2"
```
indention looks a little off

Madhu-1 replied:
Yes, Thanks for catching this one. Fixed now.

nixpanic reviewed (changes requested):
just a few nits, no real issues as far as I can spot

Madhu-1 commented:
/retest ci/centos/upgrade-tests-rbd

Madhu-1 commented:
> /retest ci/centos/upgrade-tests-rbd

`X Exiting due to GUEST_PROVISION: Failed to start host: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute
`

Madhu-1 commented:
@nixpanic manually rebased the PR and we store FsID not pool ID for cephfs volumes. Updated the examples now PTAL.

Madhu-1 commented:
/test ci/centos/upgrade-tests-cephfs

Madhu-1 commented:
> /test ci/centos/upgrade-tests-cephfs

`X Exiting due to GUEST_PROVISION: Failed to start host: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute
`

Madhu-1 commented:
@humblec @nixpanic PTAL Thanks.

Madhu-1 commented:
@ShyamsundarR @agarwal-mudit can you please help with a review here?

Madhu-1 committed:
doc: change FsID to FscID for cephfs

updated the filesystem identifier from
FsId to FscID.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 commented:
changed FsID to FscID for cephfs. @Rakshith-R Thanks for the suggestion. @Rakshith-R @nixpanic PTAL

Rakshith-R reviewed (approved):
Thanks ! Look good to me  :+1: 
(simple + easy to config + highly scale-able :100:  :wink: )

ShyamsundarR commented on internal/rbd/rbd_util.go:
```diff
@@ -1031,12 +1024,37 @@ func genVolFromVolID(
+	// Check clusterID mapping exists
+	mapping, mErr := util.GetClusterMappingInfo(vi.ClusterID)
+	if mErr != nil {
+		return vol, mErr
```
(nit) should we return err as above or the new mErr as determined here?

Madhu-1 replied:
returning `mErr` to propagate the error from GetClusterMappingInfo to the GRPC so that if there are any errors user/admin can fix them.

ShyamsundarR reviewed (approved):
One small nit, should not affect functionality though.

Madhu-1 commented:
@mergifyio rebase

Madhu-1 commented:
@mergifyio backport release-v3.4
